[Step2-3] 카드덱 구현

bluewhale0528 opened the pull request:
* 테스트코드를 먼저 짜고 추후에 테스트코드를 통과할 수 있는 코드를 작성해보는 연습을 했습니다.
   * 처음 테스트코드를 짜보는 것이어서 어떤 식으로 작성해야 할 것인가에 대해 막막했습니다. 현재 짜놓은 테스트 코드도 모든 경우를 테스트할 수 있는 코드는 아니라고 생각됩니다... 모든 경우의 수를 테스트 하기 위해서는 어떻게 작성해야 할까에 대해 고민이 많이 필요할 것 같습니다. 
* shuffle() 메소드를 구현을 위해 해당 알고리즘을 공부했습니다.
   * 랜덤으로 배열을 섞는 기능을 구현하기 위해 연결된 링크의 Shuffle 알고리즘에 대해 공부했습니다. Fisher–Yates shuffle 알고리즘 방법 중 modern method의 swap 방식을 이용해 구현했는데, 위키피디아의 예제에서는 뒤에서부터 하나씩 고정해가며 정렬하는 방식이었습니다. 하지만 저는 for문을 이용할때 인덱스가 하나씩 작아지는 것 보다는 하나씩 커지도록 구현하는 방식이 익숙해서 앞에서부터 정렬해 나가는 방식을 채택했습니다. 뒤에서부터 정렬하는 것과 앞에서부터 정렬하는 것의 차이는 없을 것 같습니다.
* removeOne() 메소드 구현을 위해 array의 remove(at: ) 속성을 이용했습니다.
   * 하나의 카드를 제거할 때 이전에 남아있던 카드 중에서 아무거나 선택해서 제거할 수 있다는 판단하에 랜덤하게 인덱스 변수 하나를 뽑아내어 그에 해당하는 값?을 리턴하면서도 원래의 카드목록에서는 제거하는 방식으로 구현했습니다.

bluewhale0528 committed:
Card 클래스 CustomStringConvertible 채택

해당 프로토콜 채택으로 description 프로퍼티 생성

bluewhale0528 committed:
CardDeck 구조체 생성

bluewhale0528 committed:
카드덱 테스트 코드 구현

테스트 코드에 필요한 메서드 틀 구현
XCTAssertNotEqual메서드 구현 위해 Card클래스 Equatable 프로토콜 채택

bluewhale0528 committed:
카드덱 초기생성 메서드 구현

bluewhale0528 committed:
테스트코드 재업로드

이전 테스트코드 구현 커밋에서 untracked된 파일로 되어있어서 커밋에 포함안되었음

bluewhale0528 committed:
shuffle 메서드 구현

bluewhale0528 committed:
removeOne 메서드 구현

bluewhale0528 committed:
reset 메서드 구현

bluewhale0528 committed:
removeOne 메서드 수정 및 관련 테스트코드 변경

godrm commented on CardGameApp/CardGameApp/Card.swift:
```diff
@@ -46,8 +46,15 @@ class Card {
         return rankString
     }
```
이 부분은 Rank 가 CustomStringConvertible 을 채택하고 구현해야 할 내용 같네요.

godrm commented on CardGameApp/CardGameApp/CardDeck.swift:
```diff
@@ -0,0 +1,53 @@
+import Foundation
+
+struct CardDeck {
+    var cards: [Card]!
```
내부 속성은 private 으로 지정하세요. 
그래야 배열을 탐색하고, 처리하는 코드가 CardDeck 내부에 있게 됩니다. 
타입 내부 속성을 가져가지 말고, 직접 처리하는 구조로 만들어보세요.

godrm commented on CardGameApp/CardGameApp/CardDeck.swift:
```diff
@@ -0,0 +1,53 @@
+            for suit in suitArray {
+                cards.append(Card(suit: Card.Suit(rawValue: suit)!, rank: Card.Rank(rawValue: count)!))
+            }
+        }
```
카드덱을 초기화할 때도 CaseIterable 프로토콜을 활용해보세요.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -28,8 +28,12 @@ class ViewController: UIViewController {
+        let heartsOfTwelve = Card(suit: .hearts, rank: .twelve)
+        print(heartsOfTwelve.description)
+        let spadesOfSeven = Card(suit: .spades, rank: .seven)
+        print(spadesOfSeven.description)
```
CustomStringConvertible 프로토콜을 채택하면 description을 접근하지 않아도 객체를 String으로 바꿔줄 수 있습니다. 참고하세요.

godrm commented on CardGameApp/CardGameAppTests/CardGameAppTests.swift:
```diff
@@ -0,0 +1,60 @@
+//
+//        let card2 = Card(suit: .hearts, rank: .one)
+//        XCTAssertEqual(card2.description, "♥️A")
+//    }
```
사용하지 않는 코드나 주석으로 처리한 코드는 과감하게 삭제하고 올리세요.

godrm commented on CardGameApp/CardGameAppTests/CardGameAppTests.swift:
```diff
@@ -0,0 +1,60 @@
+        let originCards = deck.cards
+        deck.shuffle()
+        let shuffledCards = deck.cards
+        XCTAssertNotEqual(originCards, shuffledCards)
```
shuffle 은 엄밀하게 얘기해서 기존과 동일할 수 있습니다. 
이렇게 값을 비교하는 것은 100% 완벽한 테스트 코드가 아닙니다. 
다음 단계를 하면서 어떻게 개선하면 좋은지 고민해보세요.

godrm commented:
테스트 부분 빼고 개선해서 다시 PUSH하고 댓글을 남겨주세요. 
확인하고 머지하겠습니다.

bluewhale0528 committed:
Card클래스 description 프로퍼티 재구현

bluewhale0528 committed:
CardDeck 프로퍼티 접근권한 변경

bluewhale0528 committed:
Card클래스의 Rank와 Suit CaseIterable프로토콜 채택 및 CardDeck의 createCardDeck()메소드 수정

bluewhale0528 committed:
testCode 변경

CardDeck의 프로퍼티 접근권한을 private으로 설정함으로 기존의 테스트코드 에러
XCTAssertNotEqual 메소드 실행위해 CardDeck도 Equitable프로토콜 채택하도록 변경

bluewhale0528 commented:
shuffle 테스트 코드 제외하고 코드 개선해 보았습니다. 
CardDeck구조체에 cards 프로퍼티의 접근권한을 private으로 바꿨는데, 기존 테스트 코드에서는 cards프로퍼티를 사용해서 작성했어서 해당 부분도 수정하였습니다. 
기존에 removeOne() 메소드를 테스트할 때 cards에 직접 접근해서 contains옵션으로 지워진 카드가 포함되었는지 안되었는지 알아보는 형식으로 구현했었는데, cards를 private하게 설정하니 contains옵션을 더이상 사용할 수가 없어서 그냥 메소드 실행 후 카드의 갯수를 비교하는 방식으로 변경하였습니다.

godrm commented on CardGameApp/CardGameApp/Card.swift:
```diff
@@ -29,32 +44,13 @@ class Card: CustomStringConvertible{
-    
     var description: String {
-        return "\(suit.rawValue)\(convertRankValueToString())"
+        return "\(suit.rawValue)\(rank)"
```
suit 은 여전히 CustomStringConvertible 프로토콜을 채택하지 않았네요?
enum, struct, class 모든 타입에 대한 정보를 문자열로 바꾸는 것은 CustomStringConvertible나 NSObject 를 상속받는 것으로 통일하는 게 좋습니다.

bluewhale0528 replied:
네 알겠습니다!!!
